### ISSUE TITLE
Property from a 1-buffer?

### ISSUE THREAD

**Issue by andymatuschak** [ReactiveCocoa/ReactiveCocoa#2006]
What do you folks think about this API proposal? I often find myself wanting to make signal-based APIs, but to allow for easy synchronous "sampling" in cases where that makes sense.

```swift
/// A property which makes its signal's most recently emitted value always
/// available for instantaneous access
public struct SignalProperty<T, E>: PropertyType {
  /// Initializes the property with `initialValue`, immediately starts
  /// `signalProducer` with an observing buffer of size 1. Values emitted
  /// on `signalProducer` will replace `initialValue` as the property's value.
  public init(initialValue: T, signalProducer: SignalProducer<T, E>)
}
```

**Comment by neilpa** [comment]
I've written something very similar to this but in terms of reduce which offers more flexibility. I hadn't considered wrapping it in a `PropertyType` though (I was reducing collection deltas to a collection view data source)

```swift
public struct ReducibleProperty<T, U, E>: PropertyType {
    public init(initial: U, combine: (U,T) -> U, producer: SignalProducer<T, E>)
}
```

Regardless, I'm :+1: for something like this.

**Comment by jspahrsummers** [comment]
Is this just a wrapper around `MutableProperty` plus `<~`? Not that there would necessarily be anything wrong with that, I just want to understand where it would fit in.

**Comment by andymatuschak** [comment]
Ah, yep! I hadn't noticed `<~`!

OK. Then probably the common pattern would look like this:

```
public class SomeController {
  public let someProperty: PropertyOf<Bacon> { return PropertyOf(_someProperty) }
  private let _someProperty: MutableProperty<Bacon>
  
  init(dependency: Dependency) {
    _someProperty = MutableProperty(initialValue: .Uncooked)
    _someProperty <~ dependency.baconSignalProducer
  }
}
```

My suggested API would remove some of the "ceremony" but isn't a *huge* difference:

```
public class SomeController {
  public let someProperty: PropertyOf<Bacon>

  init(dependency: Dependency) {
    someProperty = PropertyOf(SignalProperty(initialValue: .Uncooked, signalProducer: dependency.baconSignalProducer))
  }
}
```

**Comment by jspahrsummers** [comment]
I like that it's declarative, though, and expresses intention better. I'd be :+1: on adding something like that.

**Comment by andymatuschak** [comment]
Cool. I'll send a pull request after I return from honeymoon (late June).

**Comment by MichaelMcGuire** [comment]
I love this.  I find myself writing the first code block (public `PropertyOf`, private `MutableProperty`) all over the place.

**Comment by neilpa** [comment]
Resolved by #2192

**Comment by andymatuschak** [comment]
Sweet! API looks great; thanks for doing hits for me. :)
